### ISSUE TITLE
Potential bug with Microsoft.Extensions.DependencyInjection integration

### ISSUE THREAD

**Issue by UmairB** [seesharper/LightInject#357]
The following code rightly throws an error during resolve

    class Program
    {
        static void Main(string[] args)
        {
            var container = new ServiceContainer();

            container.Register<IBar, Bar>(new PerScopeLifetime());
            container.Register<IFoo, Foo>(new PerContainerLifetime());

            var foo = container.GetInstance<IFoo>();
        }
    }

    public class Foo : IFoo
    {
        public Foo(IBar bar)
        {
        }
    }

    public class Bar : IBar
    {
    }

    public interface IFoo { }

    public interface IBar { }

Due to `bar` being registered with a scope lifetime and `foo` a container lifetime. However, upon integration with `ServiceCollection` from `Microsoft.Extensions.DependencyInjection`, we have 

    var services = new ServiceCollection();
    var provider = container.CreateServiceProvider(services);

    var foo = provider.GetRequiredService<IFoo>();

And no error is thrown when getting `IFoo` from the service provider - It is resolved. I suspect this may be beyond your control as I am aware of a few issues with integration between .net core's DI and 3rd party DI libraries. If so, would you recommend that the container is verified (in say unit tests) when setting up with lightinject, and for the integration with .net core's DI just proceed as normal? Or could this have more drastic ramifications?

Tested using following project file
   <Project Sdk="Microsoft.NET.Sdk">

     <PropertyGroup>
       <OutputType>Exe</OutputType>
       <TargetFramework>netcoreapp1.1</TargetFramework>
     </PropertyGroup>

     <ItemGroup>
       <PackageReference Include="LightInject.AutoFactory" Version="2.0.0" />
       <PackageReference Include="LightInject.Microsoft.DependencyInjection" Version="2.0.1" />
       <PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="1.1.0" />
     </ItemGroup>
  
</Project>

**Comment by seesharper** [comment]
The default implementation does actually scoped services to be resolved without a current scope. This is not allowed in LightInject, but in order to pass the specification tests provided by Microsoft.Extensions.DependencyInjection.Specification.Tests, we need to allow this in the LightInject integration. 

For LightInject.Microsoft.DependencyInjection, this means starting a "root" scope that is tied to the provider instance. The IServiceProvider does not implement IDisposable and it is therefore no natural way to end this root scope. So requesting scoped services within the root scope will create a memory leak since the scope is not ended. That is just the way things are right now. 

The default implementation now adds an option to validate scopes where requesting a scoped service without a scope throws an exception. Sadly there is not specification tests for this since that is specific to the default implementation.

Bottom line, the assumption here is that this will rarely be a problem in a web application where we always have a scope that is tied to the current web request. 

As you can see in the following issue, there has been quite a lot discussion around how to handle this

https://github.com/aspnet/DependencyInjection/issues/456

**Comment by UmairB** [comment]
Ok, I understand. Thanks!
